src/lem/mod.rs

```diff
@@ -258,33 +390,81 @@ impl Func {
+                                bail!("Only tags of the same kind allowed.");
+                            }
+                        } else {
+                            kind = Some(tag_kind)
+                        }
                         if !tags.insert(tag) {
                             bail!("Tag {tag} already defined.");
                         }
                         recurse(block, return_size, map)?;
                     }
+                    match def {
+                        Some(def) => recurse(def, return_size, map)?,
+                        None => (),
+                    }
                 }
-                Ctrl::MatchSymbol(var, cases, def) => {
+                Ctrl::MatchVal(var, cases, def) => {
                     is_bound(var, map)?;
-                    let mut symbols = HashSet::new();
-                    for (symbol, block) in cases {
-                        if !symbols.insert(symbol) {
-                            bail!("Symbol {symbol} already defined.");
+                    let mut lits = HashSet::new();
+                    let mut kind = None;
+                    for (lit, block) in cases {
+                        let lit_kind = match lit {
+                            Lit::Num(..) => 0,
+                            Lit::String(..) => 1,
+                            Lit::Symbol(..) => 2,
+                        };
+                        if let Some(kind) = kind {
+                            if kind != lit_kind {
+                                bail!("Only values of the same kind allowed.");
+                            }
+                        } else {
+                            kind = Some(lit_kind)
+                        }
+                        if !lits.insert(lit) {
+                            bail!("Case {:?} already defined.", lit);
                         }
                         recurse(block, return_size, map)?;
                     }
-                    recurse(def, return_size, map)?;
+                    match def {
+                        Some(def) => recurse(def, return_size, map)?,
+                        None => (),
+                    }
+                }
+                Ctrl::IfEq(x, y, eq_block, else_block) => {
+                    is_bound(x, map)?;
+                    is_bound(y, map)?;
+                    recurse(eq_block, return_size, map)?;
+                    recurse(else_block, return_size, map)?;
                 }
             }
             Ok(())
         }
-        let map = &mut HashSet::new();
+        let map = &mut HashMap::new();
         self.input_params.iter().for_each(|var| is_unique(var, map));
-        recurse(&self.body, self.output_size, map)
+        recurse(&self.body, self.output_size, map)?;
+        for (var, u) in map.iter() {
+            let ch = var.0.chars().next().unwrap();
+            if !u && ch != '_' {
+                bail!("Variable {var} not used. If intended, please prefix it with \"_\"")
+            }
```

Comment:
Okay, this is where the `bool` used/unused bookkeeping is checked. Just a note to self, but let me know if I'm not tracking.